### PR TITLE
Add test to virtio-video

### PR DESCRIPTION
**Pull request by aesteve-rh**

### CODE REVIEW

File: tests/qtest/libqos/virtio-video.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ */
+
+#include "qemu/osdep.h"
+#include "standard-headers/linux/virtio_config.h"
+#include "../libqtest.h"
+#include "qemu/module.h"
+#include "qgraph.h"
+#include "virtio-video.h"
+
+#define NUM_VQUEUES 2
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c:20]: Indentation, suggest to have the values of the defines one below another.
Or in the IDE it looks OK?

File: tests/qtest/libqos/virtio-video.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+#define PCI_FN                  0x00
+
+static QGuestAllocator *alloc;
+
+static void virtio_video_cleanup(QVhostUserVideo *video)
+{
+    QVirtioDevice *vdev = video->vdev;
+    int i;
+
+    for (i = 0; i < NUM_VQUEUES; i++) {
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c:31]: Can you define `int i=0;` inside the `for` and remove line 29?
Here and in a few more places below.

File: tests/qtest/libqos/virtio-video.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+{
+    QVirtioDevice *vdev = video->vdev;
+    uint64_t features;
+    int i;
+
+    features = qvirtio_get_features(vdev);
+    features &= ~QVIRTIO_F_BAD_FEATURE;
+    qvirtio_set_features(vdev, features);
+
+    video->queues = g_new(QVirtQueue *, 2);
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c]: 1) `2` --> `NUM_VQUEUES`
2) Did you intend to allocate a `QVirtQueue` pointer (AFAIU every pointer is probably 8 bytes in 64 bits OS) or `QVirtQueue` itself?

File: tests/qtest/libqos/virtio-video.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    for (i = 0; i < NUM_VQUEUES; i++) {
+        video->queues[i] = qvirtqueue_setup(vdev, alloc, i);
+    }
+    qvirtio_set_driver_ok(vdev);
+}
+
+static void *qvirtio_video_get_driver(QVhostUserVideo *v_video,
+                                     const char *interface)
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c]: Looks like indentation issue (here and in few more places below)?
I mean shouldn't `c` be below `Q`?

File: tests/qtest/libqos/virtio-video.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+#include "../libqtest.h"
+#include "qemu/module.h"
+#include "qgraph.h"
+#include "virtio-video.h"
+
+#define NUM_VQUEUES 2
+#define PCI_SLOT                0x04
+#define PCI_FN                  0x00
+
+static QGuestAllocator *alloc;
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c:24]: Is there any convention for global variable names?

File: tests/qtest/libqos/virtio-video.c [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    virtio_device->obj.destructor = qvirtio_video_device_destructor;
+
+    return &virtio_device->obj;
+}
+
+/* virtio-video-pci */
+static void qvirtio_video_pci_destructor(QOSGraphObject *obj)
+{
+    QVhostUserVideoPCI *video_pci = (QVhostUserVideoPCI *) obj;
+    QOSGraphObject *pci_vobj =  &video_pci->pci_vdev.obj;
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c]: Extra space before `&video...` (here and line 122)?

File: tests/qtest/libqos/virtio-video.c [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        .extra_device_opts = "id=video0,chardev=chr-vhost-user-test",
+    };
+
+    /* vhost-user-video-device */
+    qos_node_create_driver("vhost-user-video-device",
+                            virtio_video_device_create);
+    qos_node_consumes("vhost-user-video-device", "virtio-bus", &edge_opts);
+    qos_node_produces("vhost-user-video-device", "vhost-user-video");
+
+    /* virtio-video-pci */
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.c]: Shouldn't it be `vhost-user-video-pci`?

File: tests/qtest/libqos/virtio-video.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+#ifndef TESTS_LIBQOS_VIRTIO_VIDEO_H
+#define TESTS_LIBQOS_VIRTIO_VIDEO_H
+
+#include "qgraph.h"
+#include "virtio.h"
+#include "virtio-pci.h"
+
+typedef struct QVhostUserVideo QVhostUserVideo;
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.h:20]: Why don't you use this (shorter) style to unite `typedef` and `struct` declaration? AFAIR it's possible.
```
typedef struct QVhostUserVideo {
    QVirtioDevice *vdev;
    QVirtQueue **queues;
} QVhostUserVideo;
```

File: tests/qtest/libqos/virtio-video.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    QVirtioPCIDevice pci_vdev;
+    QVhostUserVideo video;
+};
+
+struct QVhostUserVideoDevice {
+    QOSGraphObject obj;
+    QVhostUserVideo video;
+};
+
+#endif
```
Comment (barpavel) [inline tests/qtest/libqos/virtio-video.h]: I remember from my previous C experience mentioning by comment what is this `#endif` for, like:
```
#endif  // TESTS_LIBQOS_VIRTIO_VIDEO_H
```

File: tests/qtest/vhost-user-test.c [hunk 10]
```diff
@@ -1156,3 +1158,33 @@ static void register_vhost_gpio_test(void)
...
 libqos_init(register_vhost_gpio_test);
+
+static uint64_t vu_video_get_features(TestServer *s)
+{
+    return 0x1ULL << VHOST_USER_F_PROTOCOL_FEATURES;
+}
+
+static struct vhost_user_ops g_vu_video_ops = {
+    .type = VHOST_USER_VIDEO,
+
```
Comment (barpavel) [inline tests/qtest/vhost-user-test.c:1169]: Empty lines on purpose?

File: tests/qtest/libqos/virtio-video.c (hunk 2 above)
Comment (aesteve-rh) [reply tests/qtest/libqos/virtio-video.c:31]: It is customary in qemu, from what I've seen, to have all declaration at the start of the function. Even as in this case, that is the counter of a for loop and used only in that scope, it is declared at the top and initialised in the loop.

I don't love it, but I do not hate it either. And better to keep it consistent with surrounding code.

File: tests/qtest/libqos/virtio-video.c (hunk 3 above)
Comment (aesteve-rh) [reply tests/qtest/libqos/virtio-video.c]: 1. True :)
2. `video->queues` is a pointer to pointer, so it is correctly allocating a pointer.

File: tests/qtest/libqos/virtio-video.c (hunk 5 above)
Comment (aesteve-rh) [reply tests/qtest/libqos/virtio-video.c:24]: Not that I could see in the coding guidelines.

File: tests/qtest/libqos/virtio-video.h (hunk 8 above)
Comment (aesteve-rh) [reply tests/qtest/libqos/virtio-video.h:20]: It is possible, yes. This is just copying the code style of other headers in the same folder. I'll keep it as it is for consistency, although I would probably do it as you suggested.

File: tests/qtest/libqos/virtio-video.h (hunk 9 above)
Comment (aesteve-rh) [reply tests/qtest/libqos/virtio-video.h]: For small files like this one is not really that important, as you can see the define above and there is no nested defines in between. But yes, I will add it.

File: tests/qtest/vhost-user-test.c (hunk 10 above)
Comment (aesteve-rh) [reply tests/qtest/vhost-user-test.c:1169]: Yes, other `vhost_user_ops` declarations in the same file have the same empty lines.

File: tests/qtest/libqos/virtio-video.c (hunk 5 above)
Comment (barpavel) [reply tests/qtest/libqos/virtio-video.c:24]: At least there are some coding guidelines :)

**Review by barpavel — APPROVED** [review]
LGTM :)

**Pull request merged**